raysect/core/math/quaternion.pyx

```diff
@@ -0,0 +1,556 @@
+
+        ns = q1.s*q2.s - q1.x*q2.x - q1.y*q2.y - q1.z*q2.z
+        nx = q1.s*q2.x + q1.x*q2.s + q1.y*q2.z - q1.z*q2.y
+        ny = q1.s*q2.y - q1.x*q2.z + q1.y*q2.s + q1.z*q2.x
+        nz = q1.s*q2.z + q1.x*q2.y - q1.y*q2.x + q1.z*q2.s
+
+        return new_quaternion(ns, nx, ny, nz)
+
+    cdef Quaternion mul_scalar(self, double d):
+
+        cdef Quaternion q = self
+
+        return new_quaternion(d * q.s, d * q.x, d * q.y, d * q.z)
+
+    @cython.cdivision(True)
+    cpdef Quaternion inv(self):
+        """
+        Inverse operator.
+
+        .. code-block:: pycon
+
+            >>> Quaternion(0, 1, 2, 3).inv()
+            Quaternion(0.0, <-0.07143, -0.14286, -0.21429>)
+        """
+
+        cdef Quaternion q = self
+        cdef double n
+
+        n = self.norm()**2
+        return new_quaternion(q.s/n, -q.x/n, -q.y/n, -q.z/n)
+
+    cpdef double norm(self):
+        """
+        Calculates the norm of the quaternion.
+
+        .. code-block:: pycon
+
+            >>> Quaternion(0, 1, 2, 3).norm()
+            3.7416573867739413
+        """
+
+        return sqrt(self.s * self.s + self.x * self.x + self.y * self.y + self.z * self.z)
+
+    cdef Quaternion div(self, Quaternion q2):
+        """
+        Fast division operator.
+
+        This is a cython only function and is substantially faster than a call
+        to the equivalent python operator.
+        """
+
+        cdef Quaternion q1 = self, q2_inv
+
+        q2_inv = q2.inv()
+
+        return q1.mul(q2_inv)
+
+    @cython.cdivision(True)
+    cdef Quaternion div_scalar(self, double d):
+
+        cdef Quaternion q = self
+
+        # prevent divide my zero
+        if d == 0.0:
+
+            raise ZeroDivisionError("Cannot divide a quaternion by a zero scalar.")
+
+        d = 1.0 / d
+
+        return new_quaternion(d * q.s, d * q.x, d * q.y, d * q.z)
+
+    @cython.cdivision(True)
+    cpdef Quaternion normalise(self):
+        """
+        Returns a normalised copy of the quaternion.
+
+        The returned quaternion is normalised to have norm length 1.0 - a unit quaternion.
+
+        .. code-block:: pycon
+        
+            >>> a = Quaternion(0, 1, 2, 3)
+            >>> a.normalise()
+            Quaternion(0.0, <0.26726, 0.53452, 0.80178>)
+        """
+
+        cdef double n
+
+        # if current length is zero, problem is ill defined
+        n = self.norm()
+        if n == 0.0:
+
+            raise ZeroDivisionError("A zero length quaternion cannot be normalised as the direction "
+                                    "of a zero length quaternion is undefined.")
+
+        # normalise and rescale quaternion
+        n = 1.0 / n
+
+        return self.mul_scalar(n)
+
+    cpdef Quaternion copy(self):
+        """Returns a copy of this quaternion."""
+
+        return new_quaternion(self.s, self.x, self.y, self.z)
+
+    cpdef AffineMatrix3D to_transform(self):
+        """
+        Transforms the vector with the supplied AffineMatrix3D.
```

Comment:
Docstring is for the vector.transform, needs replacing. I think this method should also be called as_matrix().

Reply:
...or to_matrix() to be consistent with from_matrix() suggestion below.

Reply:
Resolved in a4d9d8a